server/channels/web/oauth.go

```diff
@@ -344,6 +345,42 @@ func completeOAuth(c *Context, w http.ResponseWriter, r *http.Request) {
 		} else { // For web
 			c.App.AttachSessionCookies(c.AppContext, w, r)
 		}
+
+		desktopToken := ""
+		if val, ok := props["desktop_token"]; ok {
+			desktopToken = val
+		}
+
+		if desktopToken != "" {
+			expiryTime := time.Now().Add(-model.DesktopTokenTTL).Unix()
+			desktopTokenErr := c.App.AuthenticateClientDesktopToken(desktopToken, expiryTime, user)
+			if desktopTokenErr != nil {
+				desktopTokenErr.Translate(c.AppContext.T)
+				c.LogErrorByCode(desktopTokenErr)
+				renderError(desktopTokenErr)
+				return
+			}
+
+			serverToken, serverTokenErr := c.App.GenerateAndSaveServerDesktopToken(desktopToken, expiryTime)
+			if serverTokenErr != nil {
+				serverTokenErr.Translate(c.AppContext.T)
+				c.LogErrorByCode(serverTokenErr)
+				renderError(serverTokenErr)
+				return
+			}
+
+			queryString := map[string]string{
+				"server_token": *serverToken,
+			}
+			if val, ok := props["redirect_to"]; ok {
+				queryString["redirect_to"] = val
+			}
+			if strings.HasPrefix(desktopToken, "dev-") {
+				queryString["isDesktopDev"] = "true"
+			}
+
+			redirectURL = utils.AppendQueryParamsToURL(c.GetSiteURLHeader()+"/login/desktop", queryString)
```

Comment:
Here we need to append the desktop client token as a query parameter too, so that the desktop can verify it against its local copy.